src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -7000,14 +6999,31 @@ public override IOperation VisitSwitchExpression(ISwitchExpressionOperation oper
             return GetCaptureReference(captureOutput, operation);
         }
 
-        private void VisitUsingVariableDeclarationOperation(IUsingDeclarationOperation operation, ImmutableArray<IOperation> statements)
+        private void VisitUsingVariableDeclarationOperation(IUsingDeclarationOperation operation, ReadOnlySpan<IOperation> statements)
         {
             IOperation? saveCurrentStatement = _currentStatement;
             _currentStatement = operation;
             StartVisitingStatement(operation);
 
-            // a using statement introduces a 'logical' block after declaration, we synthesize one here in order to analyze it like a regular using 
-            BlockOperation logicalBlock = BlockOperation.CreateTemporaryBlock(statements, ((Operation)operation).OwningSemanticModel!, operation.Syntax);
+            // a using statement introduces a 'logical' block after declaration, we synthesize one here in order to analyze it like a regular using. Don't include
+            // local functions in this block: they still belong in the containing block. We'll visit any local functions in the list after we visit the statements
+            // in this block.
+            ArrayBuilder<IOperation> statementsBuilder = ArrayBuilder<IOperation>.GetInstance();
```

Comment:
AlekseyTs commented:

> Consider allocating this builder only if there are local functions

Reply:
The only time we can avoid allocating this builder is if there are no input statements, as the array underlying the input span has the `IUsingDeclarationOperation` in it and we need to create a new block without it.

---
In reply to: [555989652](https://github.com/dotnet/roslyn/pull/50403#discussion_r555989652) [](ancestors = 555989652)